platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ProblemsProgressEventConsumer.java

```diff
@@ -52,13 +87,77 @@ public void progress(OperationIdentifier buildOperationId, OperationProgressEven
+                        problem.getException()
                     )
                 )
             );
         }
     }
 
+    private static InternalProblemCategory toInternalCategory(ProblemCategory category) {
+        return new DefaultProblemCategory(category.getNamespace(), category.getCategory(), category.getSubCategories());
+    }
+
+    private static InternalLabel toInternalLabel(String label) {
+        return new DefaultLabel(label);
+    }
+
+    private static @Nullable InternalDetails toInternalDetails(@Nullable String details) {
+        return details == null ? null : new DefaultDetails(details);
+    }
+
+    private static InternalSeverity toInternalSeverity(Severity severity) {
+        switch (severity) {
+            case ADVICE: return ADVICE;
+            case WARNING: return WARNING;
+            case ERROR: return ERROR;
+            default: return new DefaultSeverity(3); // should not happen
+        }
+    }
+
+    private static List<InternalLocation> toInternalLocations(List<ProblemLocation> locations) {
+        return locations.stream().map((Function<ProblemLocation, InternalLocation>) location -> {
+            if (location instanceof FileLocation) {
+                FileLocation fileLocation = (FileLocation) location;
+                return new DefaultFileLocation(fileLocation.getPath(), fileLocation.getLine(), fileLocation.getColumn(), fileLocation.getLength());
+            } else if (location instanceof PluginIdLocation) {
+                PluginIdLocation pluginLocation = (PluginIdLocation) location;
+                return new DefaultPluginIdLocation(pluginLocation.getPluginId());
+            } else if (location instanceof TaskPathLocation) {
+                TaskPathLocation taskLocation = (TaskPathLocation) location;
+                return new DefaultTaskPathLocation(taskLocation.getIdentityPath().toString());
+            } else {
+                throw new RuntimeException("No mapping defined for " + location.getClass().getName());
+            }
+        }).collect(Collectors.toList());
+    }
+
+    private static @Nullable InternalDocumentationLink toInternalDocumentationLink(@Nullable DocLink link) {
+        return (link == null || link.getUrl() == null) ? null : new DefaultDocumentationLink(link.getUrl());
+    }
+
+    private static List<InternalSolution> toInternalSolutions(List<String> solutions) {
+        return solutions.stream().map(DefaultSolution::new).collect(Collectors.toList());
+    }
+
+    private static InternalAdditionalData toInternalAdditionalData(Map<String, Object> additionalData) {
+        return new DefaultAdditionalData(
+            additionalData.entrySet().stream().filter(entry -> isSupportedType(entry.getValue())).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+        );
+    }
+
+    private static boolean isSupportedType(Object type) {
+        return type instanceof String || type instanceof Integer;
+    }
```

Comment:
❗Not falling inline with the JavaDoc I've seen, where it said the only acceptable type is String

Reply:
E.g.: https://github.com/gradle/gradle/pull/26991/files?diff=unified&w=0#diff-f52d08268609983db634eabdf618a41c3ea61ae922f492c6e2b32137702fc962R27

Reply:
Done